config/secretsmanager/config.go

```diff
@@ -13,5 +22,193 @@ func Configure(p *config.Provider) {
+				Region   string `json:"region"`
+			}
+
+			// traverse state
+			for _, v := range currentReplicaSet.List() {
+				// v is an interface{} type, which is a replica
+				// marshal then unmarshal to convert it to the internal type replica for easier access to fields
+				replicaBytes, err := json.JSParser.Marshal(v)
+				if err != nil {
+					continue
+				}
+				cReplica := &replica{}
+				if err := json.JSParser.Unmarshal(replicaBytes, cReplica); err != nil {
+					return nil, err
+				}
+				if cReplica.Region == "" || cReplica.KMSKeyID == "" {
+					// we should not be here, replicas at current state always have their region set and kms_key_id computed
+					continue
+				}
+
+				// add the kms_key_id to the region's kms_key_id map. For convenience, store the hash of (region,kms_key_id) pair
+				if region, ok := regionsInCurrentState[cReplica.Region]; ok {
+					region[cReplica.KMSKeyID] = hashFunc(v)
+				} else {
+					// if we are adding a kms_key_id for the region for the first time, initialize the kms_key_id map
+					regionsInCurrentState[cReplica.Region] = map[string]string{
+						cReplica.KMSKeyID: hashFunc(v),
+					}
+				}
+			}
+
+			// a convenience function for deleting diff entries for a particular replica entry
+			removeReplicaFromDiffViaHash := func(hash string) {
+				delete(diff.Attributes, fmt.Sprintf("replica.%s.kms_key_id", hash))
+				delete(diff.Attributes, fmt.Sprintf("replica.%s.region", hash))
+				delete(diff.Attributes, fmt.Sprintf("replica.%s.status", hash))
+				delete(diff.Attributes, fmt.Sprintf("replica.%s.status_message", hash))
+				delete(diff.Attributes, fmt.Sprintf("replica.%s.last_accessed_date", hash))
+			}
+			// traverse the desired Replica list at resource config (params)
+			// we want to count the Replicas, that has only region specified (no explicit kms_key_id specified, kms_key_id is left to cloud API to be automatically selected)
+			// then record this per region
+			for _, v := range desiredReplicaList {
+				// v is an interface{} type, which is a replica
+				// marshal then unmarshal to convert it to the internal type replica for easier access to fields
+				replicaBytes, err := json.JSParser.Marshal(v)
+				if err != nil {
+					continue
+				}
+				dReplica := &replica{}
+				if err := json.JSParser.Unmarshal(replicaBytes, dReplica); err != nil {
+					return nil, err
+				}
+
+				// count the region-only replicas (i.e. with automatically assigned KMS Key IDs) at parameters
+				if dReplica.KMSKeyID == "" {
+					if count, ok := desiredReplicasWithRegionOnly[dReplica.Region]; ok {
+						desiredReplicasWithRegionOnly[dReplica.Region] = count + 1
+					} else {
+						desiredReplicasWithRegionOnly[dReplica.Region] = 1
+					}
+				} else {
+					// this is a Replica at params, with explicit KMS Key ID specified in region
+					// check whether we have an exact match in current state
+					if region, ok := regionsInCurrentState[dReplica.Region]; ok {
```

Comment:
It took me a bit to figure out that `region` didn't actually have a region in it, it was just a map from kms key id -> hash. I think a different name would be better, maybe `keyHashes` or something?

Reply:
agreed, a bit verbose but maybe something like `KMSKeysWithElemHashesOfRegion`, because `keyHashes` might make one think of a hash of the key itself.

Reply:
One strategy I sometimes use for naming maps is `ValuesByKey`, so maybe something like `hashesByKMSKey`?